haskell/test/Test/Ligo/LambdaDAO.hs

```diff
@@ -0,0 +1,386 @@
+             withSender wallet1 $ call baseDao (Call @"Propose") proposeParams'
+             let proposalKey' = makeProposalKey proposeParams'
+
+             advanceToLevel (startLevel + 6 * toPeriod fs)
+             withSender wallet1 $
+               call baseDao (Call @"Vote") [PermitProtected (VoteParam proposalKey' True 200 wallet1) Nothing]
+
+             advanceToLevel (startLevel + 7 * toPeriod fs)
+             withSender admin $
+               call baseDao (Call @"Flush") (1 :: Natural)
+
+             -- Check if the result of proposal_handler_1 execution.
+             handlerStorage <- (leHandlerStorageRPC . sExtraRPC) <$> getStorageRPCLambda baseDao
+             case M.lookup [mt|key0|] handlerStorage of
+              Nothing -> fail "Expected key was not found in HandlerStorage"
+              Just x -> case lUnpackValueRaw @MText x of
+                Right ux -> assert (ux == refString) "Found unexpected value for key"
+                Left _ -> fail "Unpacking failed"
+
+      , testScenario "AddHandler proposal check checks for existing handler" $ scenario $
+         withOriginated 2
+           (\(admin: _) -> initialStorageWithExplictLambdaDAOConfig admin) $
+           \(admin:wallet1:_) fs baseDao _ -> do
+             let proposalMeta = lPackValueRaw @LambdaDaoProposalMetadata $
+                   Add_handler $ AddHandlerParam "proposal_handler_1" proposal_1_handler proposal_1_check
+             let proposalSize = metadataSize proposalMeta
+             withSender wallet1 $
+               call baseDao (Call @"Freeze") (#amount :! (proposalSize + 1000))
+             startLevel <- getOriginationLevel' @'Lambda baseDao
+
+             -- Add a new Handler
+             --
+             -- Advance one voting period to a proposing stage.
+             advanceToLevel (startLevel + toPeriod fs)
+
+             let proposeParams = ProposeParams wallet1 proposalSize proposalMeta
+             withSender wallet1 $ call baseDao (Call @"Propose") proposeParams
+
+             let proposalKey = makeProposalKey proposeParams
+
+             advanceToLevel (startLevel + 2 * toPeriod fs)
+             withSender wallet1 $
+               call baseDao (Call @"Vote") [PermitProtected (VoteParam proposalKey True 200 wallet1) Nothing]
+
+             -- Advance one voting period
+             advanceToLevel (startLevel + 3 * toPeriod fs)
+             withSender admin $
+               call baseDao (Call @"Flush") (1 :: Natural)
+
+             -- Raise a proposal to add the duplicate handler
+             advanceToLevel (startLevel + 4 * toPeriod fs)
+             let proposeParams' = ProposeParams wallet1 proposalSize proposalMeta
+             withSender wallet1 $ call baseDao (Call @"Propose") proposeParams'
+               & expectFailedWith proposalHandlerExists
+
+      , testScenario "RemoveHandler proposal check checks for existing active handler" $ scenario $
```

Comment:
I think that these should be two different tests, one that tests the failure when the handler does not exist and the other that tests it when it does exist but is no longer active.

Same for the equivalent "execute handler" test

Reply:
I thought using storage preloaded with inactive handlers cover both cases. So the error won't be thrown if  contract starts accepting non-existing handlers, or starts to accept non-active handlers, making the test fail.

But for remove handler proposal, it seems that I used storage with no handlers at all. So that was indeed a mistake.

But I have now separated the tests. May be it ll be useful in some manner.